### PR TITLE
[C++ API parity] Smooth L1 loss

### PR DESCRIPTION
**Pull request by CarMiranda**
In accordance with #25883, I added the `SmoothL1Loss` module and `smooth_l1_loss` functional.

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/functional/loss.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: nit: it would be awesome to align the indentation with the first branch of the `if` block :D

File: torch/csrc/api/include/torch/nn/functional/loss.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: nit: we could remove the new line here

File: torch/csrc/api/src/nn/modules/loss.cpp
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: It would be awesome to add a test for this as well :D

**Review by yf225 — COMMENTED** [review]
Thanks so much for the awesome work @CarMiranda! I left some minor comments.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: We can remove `SmoothL1LossOptions()` in this line to test the "no options" use case :D

File: torch/csrc/api/src/nn/modules/loss.cpp
Comment (yf225) [reply torch/csrc/api/src/nn/modules/loss.cpp]: As part of the pretty_print test, we can also test that `SmoothL1Loss(/*reduction=*/Reduction::Mean)` (i.e. the implicit constructor of `SmoothL1LossOptions`) works :D

Comment (CarMiranda) [reply torch/csrc/api/src/nn/modules/loss.cpp]: Hi @yf225, first of all, thanks for the reviews!
I had deliberately left the `pretty_print` tests out, since there were no such tests for the `L1Loss`, which has the same and unique parameter (`reduction`) as `SmoothL1Loss`, `MultiMarginLoss` and `SoftMarginLoss`. I guessed this was due to #25883, where you pointed out 

> there are still a lot of preparation work that needs to be done (e.g. fixing how we handle enums in C++ API #15149).

and these losses use the `Reduction::Reduction` enum... Should I use a `switch` statement like the one below?

```cpp
void SmoothL1LossImpl::pretty_print(std::ostream& stream) const {
  stream << "torch::nn::SmoothL1Loss(reduction=";
  switch(options.reduction()) {
    case Reduction::None:
      stream << "Reduction::None";
    break;
    case Reduction::Mean:
      stream << "Reduction::Mean";
    break;
    case Reduction::Sum:
      stream << "Reduction::Sum";
    break;
  }
  stream << ")";
}
```

File: test/cpp/api/modules.cpp
Comment (yf225) [inline test/cpp/api/modules.cpp]: We could write this as
```suggestion
  SmoothL1Loss loss(/*reduction=*/Reduction::None);
```
to also test the implicit conversion for `SmoothL1LossOptions` :D

File: torch/csrc/api/include/torch/nn/functional/loss.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: nit: would be awesome to align these two lines with the first line in `TORCH_WARN` :D

File: torch/csrc/api/src/nn/modules/loss.cpp
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: Thanks for the catch about `SmoothL1Loss` pretty printing! I checked the Python version and it seems to print `SmoothL1Loss()`, and I think we can also print the parentheses here (`L1loss` is a bad example and needs to be fixed in this regard :(

It would also be awesome to add a test for this pretty_print string, similar to all other modules :D

File: test/cpp/api/modules.cpp
Comment (CarMiranda) [reply test/cpp/api/modules.cpp]: Hi @yf225! Taking your previous reviews into account, I wrote these 4 tests for each _single-optional-parameter-loss_ module/functional pair, covering explicit instantiation (in `modules.cpp`, `*DefaultOptions`), explicit instantiation + changing parameter value (in `modules.cpp`, `*NoReduction`, this test case), omitting optional parameter for functional (in `functional.cpp`, `*DefaultOptions`) and implicit instantiation as a parameter in functional (in `functional.cpp`, `*NoReduction`).

Do you see any other test that should be run for these modules/functionals and their options? (I ask this here because you already corrected the others hehe)

**Review by yf225 — APPROVED** [review]
@CarMiranda Thanks so much for the great work! I think the tests we have now are perfect, and I will merge this after CI passes :D

**Comment by yf225** [comment]
I will rebase this on top of https://github.com/pytorch/pytorch/pull/28523 after it's merged, to get around the gcc warning problem.

**Comment by yf225** [comment]
@CarMiranda Wondering are you currently working on the RNNCell / LSTMCell / GRUCell modules? Thanks a lot for your help!

**Pull request closed without merging**
